Fair v2.0.0 with input/output in IAMC-compliant DataFrame format

JGBroadbent opened the pull request:
# Pull request

Please confirm that this pull request has done the following:

- [ ] Tests added
- [ ] Documentation added (where applicable)
- [ ] Example added (either to an existing notebook or as a new notebook, where applicable)
- [ ] Description in ``CHANGELOG.rst`` added

## Adding to CHANGELOG.rst

Please add a single line in the changelog notes similar to one of the following:

```
- (`#XX <http://link-to-pr.com>`_) Added feature which does something
- (`#XX <http://link-to-pr.com>`_) Fixed bug identified in (`#XX <http://link-to-issue.com>`_)
```

JGBroadbent committed:
Merge pull request #1 from OMS-NetZero/FAIR-v2.0.0

Merge OMS-NetZero changes onto JGBroadbent Repo

JGBroadbent committed:
FaIR 2.0 with IAMC Compliant df

Changed the input/output format of the FaIR 2.0 run functions to be IAMC Compliant DataFrames, tests to be added later.

JGBroadbent committed:
Runs using IAMC compliant DataFrames as inputs/outputs

Added in testing and IAMC compliance formatting, still needs debugging however.

JGBroadbent committed:
Update unifiedtools.py

JGBroadbent committed:
Debugged

Debugged IAMC Compliant DataFrame formatting for FaIR 2.0, passes all tests

znicholls commented on fair/ancil/units.csv:
```diff
@@ -0,0 +1,163 @@
+"Default Units for Different Variables,	"
```
can we make this csv separated rather than tab separated please?

JGBroadbent replied:
Done

znicholls commented on fair/ancil/units.py:
```diff
@@ -0,0 +1,10 @@
+
+units_filename = os.path.join(os.path.dirname(__file__), 'units.csv')
+
+class Units(dict):
```
```suggestion
class Units(dict):
    # TODO: update this so it only loads the csv once rather than once on every instantiation
```

Let's address this later

znicholls commented on fair/concentration_driven.py:
```diff
@@ -26,25 +26,15 @@ def run(inp_df, cfg):
-        "alpha": alpha_df,
-    }
-    return res_df_dict
+    res_df_iamc_compliant = unifiedtools.create_output_dataframe_aimc_compliant(    inp_df,\
```
can you format the code before the next review please, just makes it way easier to read

```suggestion
    res_df_iamc_compliant = unifiedtools.create_output_dataframe_iamc_compliant(    inp_df,\
```

znicholls commented on fair/emissions_driven.py:
```diff
@@ -10,7 +10,7 @@ def run(inp_df, cfg):
     ----------
     inp_df : :obj:`pd.DataFrame`
-        Input :obj:`pd.DataFrame` containing the timeseries to run
+        Input :obj:`pd.DataFrame` containing the timeseries to run, in IAMC compliant DataFrame format (i.e. A multiIndex of Model, Region, Scenario, Unit, Variable then Columns for time)
```
Is the aim for it to be IAMC compliant or simply useable by https://github.com/IAMconsortium/pyam?

JGBroadbent replied:
I figured I might as well make the output IAMC compliant, but you're right that actually it makes rather more sense to have input as useable rather than explicitly compliant (but I'm guessing the code doesn't need to change at all

znicholls commented on fair/tools/unifiedtools.py:
```diff
@@ -1,6 +1,9 @@
 import pandas as pd
+import pyam as pyam
+
+from ..ancil.units import *
```
```suggestion
from ..ancil.units import Units
```

Please don't ever use `*` imports, they're a nightmare

JGBroadbent replied:
Fair enough!

znicholls commented on fair/tools/unifiedtools.py:
```diff
@@ -196,6 +199,86 @@ def create_output_dataframes(inp_df, gas_np, RF_np, T_np, alpha_np, ext_forcing_
+
+    units = Units()
+
+    variable_array = inp_df.timeseries().index.levels[3].to_numpy()
```
```suggestion
    inp_df_ts = inp_df.timeseries()
    variable_index_level = inp_df_ts.index.names.tolist().index("variable")
    variable_array = inp_df_ts.index.levels[variable_index_level].to_numpy()
```

I think something like that is safer and easier to read rather than having this magic number 3 just appear

JGBroadbent replied:
Yes, good point, I've tried to avoid this elsewhere

znicholls commented on fair/tools/unifiedtools.py:
```diff
@@ -196,6 +199,86 @@ def create_output_dataframes(inp_df, gas_np, RF_np, T_np, alpha_np, ext_forcing_
+    split_variable_array = np.array([variable_array[i].split('|',1) for i in range(len(variable_array))])
+
+    gas_array = split_variable_array[:,1]
+    function_array = split_variable_array[:,0]
```
what do you mean by function array? This is an array of strings, not functions?

znicholls commented on fair/tools/unifiedtools.py:
```diff
@@ -196,6 +199,86 @@ def create_output_dataframes(inp_df, gas_np, RF_np, T_np, alpha_np, ext_forcing_
+
+    model_region_scenario_array = np.unique(inp_df.timeseries().index.droplevel(('unit','variable')).to_numpy())
+    if len(model_region_scenario_array) > 1:
+        raise Exception('Error: More than one Model, Region + Scenario combination input passed')
```
```suggestion
        raise ValueError('More than one Model, Region + Scenario combination input passed')
```

??

Something more specific than the very general Exception would be great

znicholls commented on fair/tools/unifiedtools.py:
```diff
@@ -196,6 +199,86 @@ def create_output_dataframes(inp_df, gas_np, RF_np, T_np, alpha_np, ext_forcing_
+    time_index = inp_df.timeseries().columns
+
+    data_array = np.array([[]])
+    for i in range(len(gas_array)):
```
```suggestion
    for i, gas_name in enumerate(gas_array):
```

znicholls commented on fair/tools/unifiedtools.py:
```diff
@@ -196,6 +199,86 @@ def create_output_dataframes(inp_df, gas_np, RF_np, T_np, alpha_np, ext_forcing_
+
+    data_array = np.array([[]])
+    for i in range(len(gas_array)):
+        gas_name = gas_array[i]
```
```suggestion
```

znicholls commented on fair/tools/unifiedtools.py:
```diff
@@ -196,6 +199,86 @@ def create_output_dataframes(inp_df, gas_np, RF_np, T_np, alpha_np, ext_forcing_
+        data_array = np.append(data_array,[RF_output_array],axis=0)
+        data_array = np.append(data_array,[alpha_output_array],axis=0)
+
+    ext_RF_output_variable = "Effective Radiative Forcing|External Forcing"
```
```suggestion
    ext_RF_output_variable = "Effective Radiative Forcing|Other"
```

or "Prescribed"? External Forcing doesn't make any sense to me and will confuse people I suspect

znicholls commented on fair/tools/unifiedtools.py:
```diff
@@ -196,6 +199,86 @@ def create_output_dataframes(inp_df, gas_np, RF_np, T_np, alpha_np, ext_forcing_
+
+    pyam_df = pyam.IamDataFrame(output_df)
+
+    output_df = pyam.IamDataFrame(pyam_df.append(inp_df))
```
It'll be quicker to append using pandas than pyam i.e. you only want to call `pyam.IamDataFrame` once (rather than twice as it is now)

znicholls commented on fair/tools/unifiedtools.py:
```diff
@@ -229,13 +312,15 @@ def return_np_function_arg_list(inp_df, cfg, concentration_mode=False):
 
-    inp_df = inp_df.iloc[inp_df.index.astype("str").str.lower().argsort()]
-    time_index = inp_df.index
+    time_index = inp_df.timeseries().columns
```
You might want to check you've got integers here. Some pyam dataframes have funky time axes so a quick assert is worth it

znicholls replied:
Hang on, what type is `inp_df`? If it's an `ScmRun`, you can just do `inp_df.timeseries(time_axis="year")` and then you'll get years as timesteps rather than pandas timestamps

znicholls commented on fair/tools/unifiedtools.py:
```diff
@@ -229,13 +312,15 @@ def return_np_function_arg_list(inp_df, cfg, concentration_mode=False):
+    time_index = inp_df.timeseries().columns
 
-    timestep_np = np.append(np.diff(time_index), np.diff(time_index)[-1])
+    #Timestep is given in nanoseconds initially
```
this shouldn't be needed

znicholls commented on pytest-env/bin/Activate.ps1:
```diff
@@ -0,0 +1,241 @@
+<#
```
this file shouldn't be in here

znicholls commented on pytest-env/bin/activate:
```diff
@@ -0,0 +1,76 @@
+# This file must be used with "source bin/activate" *from bash*
```
shouldn't be here

znicholls commented on pytest-env/bin/activate.csh:
```diff
@@ -0,0 +1,37 @@
+# This file must be used with "source bin/activate.csh" *from csh*.
```
shouldn't be here

similarly for everything else in the pytest-env folder

JGBroadbent committed:
speed improvements + removed old formatting

JGBroadbent committed:
Delete pyvenv.cfg

JGBroadbent committed:
Delete python3

JGBroadbent committed:
Delete python

JGBroadbent committed:
Delete pytest

JGBroadbent committed:
Delete py.test

JGBroadbent committed:
Delete pip3.8

JGBroadbent committed:
Delete pip3

JGBroadbent committed:
Delete pip

JGBroadbent committed:
Delete pint-convert

JGBroadbent committed:
Delete f2py3.8

JGBroadbent committed:
Delete f2py3

JGBroadbent committed:
Delete easy_install-3.8

JGBroadbent committed:
Delete easy_install

JGBroadbent committed:
Delete f2py

JGBroadbent committed:
Delete chardetect

JGBroadbent committed:
Delete activate.fish

JGBroadbent committed:
Delete activate.csh

JGBroadbent committed:
Delete activate

JGBroadbent committed:
Delete Activate.ps1

JGBroadbent committed:
Update units to be csv not tab separated

JGBroadbent committed:
Update units.py

JGBroadbent committed:
Update fair/concentration_driven.py

Co-authored-by: Zeb Nicholls <zebedee.nicholls@climate-energy-college.org>

JGBroadbent committed:
Update fair/tools/unifiedtools.py

Co-authored-by: Zeb Nicholls <zebedee.nicholls@climate-energy-college.org>

JGBroadbent committed:
Update fair/tools/unifiedtools.py

Co-authored-by: Zeb Nicholls <zebedee.nicholls@climate-energy-college.org>

JGBroadbent committed:
Update fair/tools/unifiedtools.py

Co-authored-by: Zeb Nicholls <zebedee.nicholls@climate-energy-college.org>

JGBroadbent committed:
Update External Forcing -> Other

JGBroadbent committed:
Merge branch 'FAIR-v2.0.0' of https://github.com/JGBroadbent/FAIR into FAIR-v2.0.0

JGBroadbent committed:
change name of iamc compliant

JGBroadbent committed:
pep8 compliance

JGBroadbent committed:
Add in default parameter capability

chrisroadmap committed:
add pyam-iamc dependency

JGBroadbent committed:
Update default_gas_parameters.csv

chrisroadmap committed:
PEP up a bit and pepper with docstrings

chrisroadmap committed:
update changelog

chrisroadmap committed:
format

chrisroadmap committed:
black

chrisroadmap committed:
tidy up imports

JGBroadbent committed:
Added in constants

JGBroadbent committed:
Merge pull request #2 from chrisroadmap/fix-install

Add the pyam-iamc dependency in setup

JGBroadbent committed:
get rid of pytest env again

JGBroadbent committed:
try a new pytest_project

JGBroadbent committed:
try again

JGBroadbent committed:
removed pytest_project from tracked files

JGBroadbent committed:
.

JGBroadbent committed:
removed pytest_project from tracked files

JGBroadbent committed:
Update .gitignore

JGBroadbent committed:
Update .gitignore

JGBroadbent committed:
test gitignore

JGBroadbent committed:
Merge branch 'FAIR-v2.0.0' of https://github.com/JGBroadbent/FAIR into FAIR-v2.0.0

JGBroadbent committed:
removed pytest_project from tracked files

JGBroadbent committed:
test

JGBroadbent committed:
removed pytest_project from tracked files

JGBroadbent committed:
test

JGBroadbent committed:
Updated to support secondary forcings

Added in functionality to support secondary forcing in the format Gas|Secondary Forcing, with tests etc

JGBroadbent committed:
Add support for secondary sources of the same gas

Added in support for using multiple instances of the same gas, e.g. 'CO2', 'CO2|Transport','CO2|Power' which might all want to use the default CO2 parameter set

JGBroadbent committed:
Include Forcing as part of IAMC compliant input + cleanup

Added in Forcing as part of IAMC input + cleaned up the code.

JGBroadbent committed:
Cleaned up unifiedtools.py

JGBroadbent committed:
Adding in 1.6 interpretation

JGBroadbent committed:
.

JGBroadbent committed:
Add in proper formatting for 1.6 interpretter

JGBroadbent committed:
Update one_point_six_interpretter.py

JGBroadbent committed:
Update Doc Strings

JGBroadbent committed:
Example Notebook Added

JGBroadbent committed:
Added in RCP defaults and started work on Example Notebook

JGBroadbent committed:
Update Example Notebook

JGBroadbent committed:
Update Example-Usage.ipynb

chrisroadmap commented:
I'm going to merge this into the main repo now just to make developing a bit easier. Fork of a fork isn't working for me
